### PR TITLE
[Low CPU memory] + device map

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Co-Authored-By: Pi Esposito [piero.skywalker@gmail.com](mailto:piero.skywalker@gmail.com)

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/safety_checker.py [hunk 1]
```diff
@@ -19,6 +19,9 @@ def cosine_distance(image_embeds, text_embeds):
 class StableDiffusionSafetyChecker(PreTrainedModel):
     config_class = CLIPConfig
 
+    _no_split_modules = ["CLIPEncoderLayer"]
+    base_model_prefix = "vision_model"
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/safety_checker.py:23]:
```suggestion
```

File: src/diffusers/pipelines/stable_diffusion/safety_checker.py [hunk 2]
```diff
@@ -19,6 +19,8 @@ def cosine_distance(image_embeds, text_embeds):
 class StableDiffusionSafetyChecker(PreTrainedModel):
     config_class = CLIPConfig
 
+    _no_split_modules = ["CLIPEncoderLayer"]
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/safety_checker.py:22]: needed to allow to use `device_map="auto"`

**Comment by patrickvonplaten** [comment]
This PR allows the following to be fully functional once the two following are merged:

- https://github.com/huggingface/accelerate/pull/747
- https://github.com/huggingface/transformers/commit/06514b3e1a00f031ca5bf999a87081fd1fe8b700

**Comment by Thomas-MMJ** [comment]
looks like isort needs to be ran to make the linter happy.

File: tests/test_pipelines.py [hunk 3]
```diff
@@ -17,12 +17,15 @@
 import os
 import random
 import tempfile
+import tracemalloc
 import unittest
 
 import numpy as np
 import torch
 
+import accelerate
```
Comment (Thomas-MMJ) [inline tests/test_pipelines.py:26]: Need to sort the imports to make the linter happy, running isort locally gives this order,

```
import os
import random
import tempfile
import tracemalloc
import unittest

import accelerate
import numpy as np
import PIL
import torch

import transformers
```

File: tests/test_pipelines.py [hunk 4]
```diff
@@ -17,12 +17,15 @@
...
 import PIL
+import transformers
```
Comment (Thomas-MMJ) [inline tests/test_pipelines.py:28]: see above comment for isort import grouping and ordering.

File: tests/test_pipelines.py [hunk 5]
```diff
@@ -49,6 +52,7 @@
 from diffusers.pipeline_utils import DiffusionPipeline
 from diffusers.schedulers.scheduling_utils import SCHEDULER_CONFIG_NAME
 from diffusers.utils import CONFIG_NAME, WEIGHTS_NAME, floats_tensor, load_image, slow, torch_device
+from packaging import version
```
Comment (Thomas-MMJ) [inline tests/test_pipelines.py:55]: might also need to check with isort, locally it makes changes to their original commit also (makes the multiple item imports into a multiline tuple, and sorts PIL to the top), so my isort might be different than their config.

**Comment by patrickvonplaten** [comment]
PRs on `transformers` and `accelerate` are merged! `device_map="auto"` should now work if using both `transformers` and `accelerate` on master. 

Think we can merge here. @piEsposito, again very sorry about fiddeling with your PR so much - ok if we merge this PR (you'll be a co-author of it)?

**Comment by piEsposito** [comment]
@patrickvonplaten no problem at all, things like that happen all the time. I really appreciate how you handled it and your kindness of keeping me as a co-author. 

IMO you can merge this.

**Pull request merged**